middle_end/closure/closure.ml

```diff
@@ -697,15 +697,39 @@ let rec substitute loc ((backend, fpc) as st) sb rn ulam =
   | Uunreachable ->
       Uunreachable
 
-(* Perform an inline expansion *)
+(* Perform an inline expansion:
 
+   If [f p = body], substitute [f a] by [let p = a in body].
+
+   Under certain conditions, further simplifications are possible (we use the
+   terminology of [Semantics_of_primitives], applied to terms of the Clambda
+   language):
+
+   - [f a] is equivalent to [body[a/p]] if [a] has no effects and no coeffects.
+     However, we only want to do this rewriting if [body[a/p]] does not increase
+     the size of [body]. Since this is hard to decide in general, as an
+     approximation, only consider the case when [a] is an immutable variable or
+     a constant.
+
+   - [f a] is equivalent to [body] if [p] does not occur in [body] and [a] has
+     only generative effects. As an approximation we only do this when [a] is
+     either a variable or a constant or a closure allocation.
+
+   - In general [f a] is equivalent to [a; body] if [p] does not occur in [body].
+*)
+
+let mutable_ids = Hashtbl.create 17
+
+(* Approximates "no effects and no coeffects" *)
 let is_simple_argument = function
-  | Uvar _  | Uconst _ -> true
+  | Uvar v -> not (Hashtbl.mem mutable_ids v)
+  | Uconst _ -> true
   | _ -> false
 
+(* Approximates "only generative effects" *)
 let no_effects = function
-  | Uclosure _ -> true
-  | u -> is_pure u
+  | Uclosure _ | Uvar _ | Uconst _ -> true
+  | _ -> false
```

Comment:
In several cases your patch replaces `is_pure u` by `false` (for example, "pure" primitives, offset instructions, etc., will now be rejected while they were previously accepted). This looks like a regression in optimizations, am I missing something? If I understand your comment correctly, this function is used to decide whether a subexpression can be dropped completely, so we shouldn't need to ensure that accepted expressions are "small".

Reply:
Oops, indeed, I got confused here. I reverted this change.